ssh/tailssh/tailssh.go

```diff
@@ -167,7 +165,29 @@ func (srv *server) handleSSH(s ssh.Session) {
 		s.Exit(1)
 	}
 
-	logf("ssh: connection from %v %v to %v@ => %q. command = %q, env = %q", srcIP, uprof.LoginName, sshUser, localUser, s.Command(), s.Environ())
+	var sessionClose chan error
+	if action.SesssionDuration != 0 {
+		sessionClose = make(chan error, 1)
+		t := time.AfterFunc(action.SesssionDuration, func() {
+			sessionClose <- fmt.Errorf("max duration %v elapsed", action.SesssionDuration)
+		})
+		defer t.Stop()
+	}
+	srv.handleAcceptedSSH(s, sessionClose, sctx, lu)
+}
+
+// handleAcceptedSSH handles s once it's been accepted and determined
+// that it should run as user lu.
+//
+// sessionClose, if non-nil, is readable when the session should
+// be forcefully closed with an error that says why.
+func (srv *server) handleAcceptedSSH(s ssh.Session, sessionClose <-chan error, sctx *sshContext, lu *user.User) {
```

Comment:
use a context with timeout/cancel? you lose the error but that's more canonical?

Reply:
or you could embed https://pkg.go.dev/context#Context in `sshContext`, make it also expose a `CancelWithErr` function and then have it override the `ctx.Err()` to instead return the err from cancel.

`context.WithTimeoutError`/`context.WithCancelError` might be something I want in general tbh, would be useful to figure out things like who canceled the context and why.